--- lirc-0.9.3/lib/Makefile.am.orig	2015-09-02 21:43:49.000000000 +0200
+++ lirc-0.9.3/lib/Makefile.am	2015-10-06 16:26:21.296067036 +0200
@@ -31,6 +31,7 @@
                               transmit.c
 
 libirrecord_la_LIBADD       = liblirc.la
+libirrecord_la_LIBADD = liblirc.la
 libirrecord_la_SOURCES      = irrecord.c
 
 liblirc_client_la_LDFLAGS   = -version-info 6:0:6
@@ -49,6 +50,7 @@
                               serial.c \
                               transmit.c \
                               transmit.h
+liblirc_driver_la_LIBADD = liblirc.la
 
 dist_include_HEADERS        = lirc_client.h \
                               lirc_driver.h \
