--- lirc-0.9.0/./drivers/lirc_atiusb/lirc_atiusb.c~	2011-03-25 23:28:18.000000000 +0100
+++ lirc-0.9.0/./drivers/lirc_atiusb/lirc_atiusb.c	2011-09-10 11:19:50.422542829 +0200
@@ -48,7 +48,6 @@
 #include <linux/slab.h>
 #include <linux/module.h>
 #include <linux/kmod.h>
-#include <linux/smp_lock.h>
 #include <linux/completion.h>
 #include <linux/uaccess.h>
 #include <linux/usb.h>
--- lirc-0.9.0/drivers/lirc_i2c/lirc_i2c.c~	2011-03-25 23:28:18.000000000 +0100
+++ lirc-0.9.0/drivers/lirc_i2c/lirc_i2c.c	2011-09-10 12:24:55.710859996 +0200
@@ -555,8 +555,8 @@
 		kfree(ir);
 		return -EINVAL;
 	}
-	printk(KERN_INFO "lirc_i2c: chip 0x%x found @ 0x%02x (%s)\n",
-	       adap->id, addr, ir->c.name);
+	printk(KERN_INFO "lirc_i2c: chip found @ 0x%02x (%s)\n",
+	       addr, ir->c.name);
 
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 31)
 	/* register device */
--- lirc-0.9.0/drivers/lirc_gpio/lirc_gpio.c.orig	2011-03-25 23:28:18.000000000 +0100
+++ lirc-0.9.0/drivers/lirc_gpio/lirc_gpio.c	2011-09-10 16:30:47.051485373 +0200
@@ -38,7 +38,6 @@
 #include <linux/errno.h>
 
 #include "../drivers/media/video/bt8xx/bttv.h"
-#include "../drivers/media/video/bt8xx/bttvp.h"
 
 #include "drivers/kcompat.h"
 #if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 35)
--- lirc-0.9.0/drivers/lirc_serial/lirc_serial.c~	2011-03-25 23:28:18.000000000 +0100
+++ lirc-0.9.0/drivers/lirc_serial/lirc_serial.c	2012-07-01 14:29:13.757144222 +0200
@@ -48,6 +48,7 @@
  * Steve Davies <steve@daviesfam.org>  July 2001
  */
 
+#include <linux/version.h>
 #include <linux/module.h>
 #include <linux/errno.h>
 #include <linux/signal.h>
@@ -66,7 +66,9 @@
 #include <linux/poll.h>
 #include <linux/platform_device.h>
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(3, 4, 0)
 #include <asm/system.h>
+#endif
 #include <linux/io.h>
 #include <linux/irq.h>
 #include <linux/fcntl.h>
--- lirc-0.9.0/drivers/lirc_sir/lirc_sir.c~	2011-03-25 23:28:18.000000000 +0100
+++ lirc-0.9.0/drivers/lirc_sir/lirc_sir.c	2012-07-01 14:29:56.867142704 +0200
@@ -59,7 +59,9 @@
 #include <linux/mm.h>
 #include <linux/delay.h>
 #include <linux/poll.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(3, 4, 0)
 #include <asm/system.h>
+#endif
 #include <linux/io.h>
 #include <asm/irq.h>
 #include <linux/fcntl.h>
--- lirc-0.9.0/drivers/lirc_imon/lirc_imon.c~	2011-03-25 23:28:18.000000000 +0100
+++ lirc-0.9.0/drivers/lirc_imon/lirc_imon.c	2012-09-25 22:50:43.426536543 +0200
@@ -61,6 +61,11 @@
 
 #define BIT_DURATION	250	/* each bit received is 250us */
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 5, 0)
+#define err(format, arg...)					\
+	printk(KERN_ERR KBUILD_MODNAME ": " format "\n", ##arg)
+#endif
+
 /*** P R O T O T Y P E S ***/
 
 /* USB Callback prototypes */
--- lirc-0.9.0/drivers/lirc_sasem/lirc_sasem.c~	2011-03-25 23:28:18.000000000 +0100
+++ lirc-0.9.0/drivers/lirc_sasem/lirc_sasem.c	2012-09-25 22:51:20.226535245 +0200
@@ -73,6 +73,11 @@
 
 #define IOCTL_LCD_CONTRAST 1
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 5, 0)
+#define err(format, arg...)					\
+	printk(KERN_ERR KBUILD_MODNAME ": " format "\n", ##arg)
+#endif
+
 /*** P R O T O T Y P E S ***/
 
 /* USB Callback prototypes */
--- lirc-0.9.0/drivers/lirc_ttusbir/lirc_ttusbir.c~	2011-03-25 23:28:18.000000000 +0100
+++ lirc-0.9.0/drivers/lirc_ttusbir/lirc_ttusbir.c	2012-09-25 22:51:38.126534613 +0200
@@ -58,6 +58,11 @@
 #define DPRINTK(_x_, a...)
 #endif
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 5, 0)
+#define err(format, arg...)					\
+	printk(KERN_ERR KBUILD_MODNAME ": " format "\n", ##arg)
+#endif
+
 /* function declarations */
 static int probe(struct usb_interface *intf, const struct usb_device_id *id);
 static void disconnect(struct usb_interface *intf);
