--- lirc-0.8.6/config.h.in.orig	2009-09-13 10:52:17.000000000 +0200
+++ lirc-0.8.6/config.h.in	2009-09-17 21:06:09.662437650 +0200
@@ -348,7 +348,7 @@
 
 #define LIRC_IRTTY		DEVDIR "/" "ttyS0"
 
-#define LIRCD			VARRUNDIR "/" PACKAGE "/" DEV_LIRCD
+#define LIRCD			"/tmp/.lircd"
 #define LIRCM			VARRUNDIR "/" PACKAGE "/" DEV_LIRCM
 
 #define LIRCDCFGFILE		SYSCONFDIR "/" PACKAGE "/" CFG_LIRCD
--- lirc-0.8.6/configure.ac.orig	2009-09-17 21:04:55.172445270 +0200
+++ lirc-0.8.6/configure.ac	2009-09-17 21:06:45.876392377 +0200
@@ -1742,7 +1742,7 @@
 
 #define LIRC_IRTTY		DEVDIR "/" "ttyS0"
 
-#define LIRCD			VARRUNDIR "/" PACKAGE "/" DEV_LIRCD
+#define LIRCD			"/tmp/.lircd"
 #define LIRCM			VARRUNDIR "/" PACKAGE "/" DEV_LIRCM
 
 #define LIRCDCFGFILE		SYSCONFDIR "/" PACKAGE "/" CFG_LIRCD
