--- lirc-0.8.3/config.h.in.orig	2008-05-04 15:59:49.000000000 +0200
+++ lirc-0.8.3/config.h.in	2008-05-05 18:39:25.000000000 +0200
@@ -342,7 +342,7 @@
 
 #define LIRC_IRTTY		DEVDIR "/" "ttyS0"
 
-#define LIRCD			DEVDIR "/" DEV_LIRCD
+#define LIRCD			"/tmp/.lircd"
 #define LIRCM			DEVDIR "/" DEV_LIRCM
 
 #define LIRCDCFGFILE		SYSCONFDIR "/" CFG_LIRCD
--- lirc-0.8.3/configure.ac.orig	2008-05-05 18:37:21.000000000 +0200
+++ lirc-0.8.3/configure.ac	2008-05-05 18:39:14.000000000 +0200
@@ -1664,7 +1664,7 @@
 
 #define LIRC_IRTTY		DEVDIR "/" "ttyS0"
 
-#define LIRCD			DEVDIR "/" DEV_LIRCD
+#define LIRCD			"/tmp/.lircd"
 #define LIRCM			DEVDIR "/" DEV_LIRCM
 
 #define LIRCDCFGFILE		SYSCONFDIR "/" CFG_LIRCD
