diff -urN lirc-0.6.3.org/daemons/Makefile.am lirc-0.6.3/daemons/Makefile.am
--- lirc-0.6.3.org/daemons/Makefile.am	Thu Jul 12 00:41:24 2001
+++ lirc-0.6.3/daemons/Makefile.am	Thu Jul 12 01:54:47 2001
@@ -73,12 +73,10 @@
 uninstall-local: rmdev
 
 install-data-local:
-	test "$(lircd_conf)" = ""  || test -e $(sysconfdir)/lircd.conf  || ($(mkinstalldirs) $(sysconfdir) && $(INSTALL_DATA) $(top_srcdir)/remotes/$(lircd_conf) $(sysconfdir)/lircd.conf)
-	test "$(lircmd_conf)" = "" || test -e $(sysconfdir)/lircmd.conf || ($(mkinstalldirs) $(sysconfdir) && $(INSTALL_DATA) $(top_srcdir)/remotes/$(lircmd_conf) $(sysconfdir)/lircmd.conf)
+	test "$(lircd_conf)" = ""  || test -e $(DESTDIR)$(sysconfdir)/lircd.conf  || ($(mkinstalldirs) $(DESTDIR)$(sysconfdir) && $(INSTALL_DATA) $(top_srcdir)/remotes/$(lircd_conf) $(DESTDIR)$(sysconfdir)/lircd.conf)
+	test "$(lircmd_conf)" = "" || test -e $(DESTDIR)$(sysconfdir)/lircmd.conf || ($(mkinstalldirs) $(DESTDIR)$(sysconfdir) && $(INSTALL_DATA) $(top_srcdir)/remotes/$(lircmd_conf) $(DESTDIR)$(sysconfdir)/lircmd.conf)
 
 mkdev:
-	test -e @devdir@/lircd || @mknod@ @devdir@/lircd p
-	test -e @devdir@/lircm || @mknod@ @devdir@/lircm p
 
 rmdev:
 	$(RM) @devdir@/lircd
diff -urN lirc-0.6.3.org/drivers/lirc_dev/Makefile.am lirc-0.6.3/drivers/lirc_dev/Makefile.am
--- lirc-0.6.3.org/drivers/lirc_dev/Makefile.am	Thu Jul 12 00:41:24 2001
+++ lirc-0.6.3/drivers/lirc_dev/Makefile.am	Thu Jul 12 01:52:06 2001
@@ -30,7 +30,6 @@
 uninstall-local: rmdev
 
 mkdev:
-	test -c @devdir@/lirc || @mknod@ @devdir@/lirc c @lirc_major@ 0
 
 rmdev:
 	test -c @devdir@/lirc && $(RM) @devdir@/lirc
diff -urN lirc-0.6.3.org/drivers/lirc_gpio/Makefile.am lirc-0.6.3/drivers/lirc_gpio/Makefile.am
--- lirc-0.6.3.org/drivers/lirc_gpio/Makefile.am	Thu Jul 12 00:41:24 2001
+++ lirc-0.6.3/drivers/lirc_gpio/Makefile.am	Thu Jul 12 01:52:06 2001
@@ -30,7 +30,6 @@
 uninstall-local: rmdev
 
 mkdev:
-	test -c @devdir@/lirc || @mknod@ @devdir@/lirc c @lirc_major@ 0
 
 rmdev:
 	test -c @devdir@/lirc && $(RM) @devdir@/lirc
diff -urN lirc-0.6.3.org/drivers/lirc_i2c/Makefile.am lirc-0.6.3/drivers/lirc_i2c/Makefile.am
--- lirc-0.6.3.org/drivers/lirc_i2c/Makefile.am	Thu Jul 12 00:41:24 2001
+++ lirc-0.6.3/drivers/lirc_i2c/Makefile.am	Thu Jul 12 01:52:06 2001
@@ -32,7 +32,6 @@
 uninstall-local: rmdev
 
 mkdev:
-	test -c @devdir@/lirc || @mknod@ @devdir@/lirc c @lirc_major@ 0
 
 rmdev:
 	test -c @devdir@/lirc && $(RM) @devdir@/lirc
diff -urN lirc-0.6.3.org/drivers/lirc_parallel/Makefile.am lirc-0.6.3/drivers/lirc_parallel/Makefile.am
--- lirc-0.6.3.org/drivers/lirc_parallel/Makefile.am	Thu Jul 12 00:41:24 2001
+++ lirc-0.6.3/drivers/lirc_parallel/Makefile.am	Thu Jul 12 01:52:06 2001
@@ -30,7 +30,6 @@
 uninstall-local: rmdev
 
 mkdev:
-	test -c @devdir@/lirc || @mknod@ @devdir@/lirc c @lirc_major@ 0
 
 rmdev:
 	test -c @devdir@/lirc && $(RM) @devdir@/lirc
diff -urN lirc-0.6.3.org/drivers/lirc_serial/Makefile.am lirc-0.6.3/drivers/lirc_serial/Makefile.am
--- lirc-0.6.3.org/drivers/lirc_serial/Makefile.am	Thu Jul 12 00:41:24 2001
+++ lirc-0.6.3/drivers/lirc_serial/Makefile.am	Thu Jul 12 01:52:06 2001
@@ -31,7 +31,6 @@
 uninstall-local: rmdev
 
 mkdev:
-	test -c @devdir@/lirc || @mknod@ @devdir@/lirc c @lirc_major@ 0
 
 rmdev:
 	test -c @devdir@/lirc && $(RM) @devdir@/lirc
diff -urN lirc-0.6.3.org/drivers/lirc_sir/Makefile.am lirc-0.6.3/drivers/lirc_sir/Makefile.am
--- lirc-0.6.3.org/drivers/lirc_sir/Makefile.am	Thu Jul 12 00:41:24 2001
+++ lirc-0.6.3/drivers/lirc_sir/Makefile.am	Thu Jul 12 01:52:06 2001
@@ -30,7 +30,6 @@
 uninstall-local: rmdev
 
 mkdev:
-	test -c @devdir@/lirc || @mknod@ @devdir@/lirc c @lirc_major@ 0
 
 rmdev:
 	test -c @devdir@/lirc && $(RM) @devdir@/lirc
