diff -ruN lirc-0.6.6.old/configure.in lirc-0.6.6/configure.in
--- lirc-0.6.6.old/configure.in	Sun Oct  6 11:21:08 2002
+++ lirc-0.6.6/configure.in	Sat Mar  1 12:38:52 2003
@@ -94,7 +94,7 @@
 AC_MSG_RESULT()
 
 possible_drivers="(none) (udp) (dsp) (remotemaster) (pctv) (logitech) \
-	(slinke) (silitek) (mediafocusI) (irreal)"
+	(makpc) (slinke) (silitek) (mediafocusI) (irreal)"
 
 if test x${no_kernel} != xyes; then
   possible_drivers="${possible_drivers} (lirc_serial) (lirc_sir) \
@@ -124,7 +124,7 @@
 			  chronos, comX, dsp, cph03x, cph06x,
 			  creative, fly98, flyvideo, hauppauge,
 			  hauppauge_dvb, irdeo, irdeo_remote, irman,
-			  irreal, it87, knc_one, logitech, lptX,
+			  irreal, it87, knc_one, logitech, lptX, makpc,
 			  mediafocusI, packard_bell, parallel, pctv,
 			  pixelview_bt878, pixelview_pak,
 			  pixelview_pro, provideo, realmagic,
@@ -172,6 +172,9 @@
 	  logitech)
 	  hw_module="${hw_module} hw_logitech.o serial.o"
 	  ;;
+	  makpc)
+	  hw_module="${hw_module} hw_makpc.o serial.o"
+	  ;;
 	  slinke)
 	  hw_module="${hw_module} hw_slinke.o serial.o receive.o"
 	  ;;
@@ -432,6 +435,13 @@
   lircd_conf="logitech/lircd.conf.logitech"
   lircmd_conf="logitech/lircmd.conf.logitech"
 fi
+if test "$driver" = "makpc"; then
+  lirc_driver="none"
+  hw_module="hw_makpc.o serial.o"
+  HW_DEFAULT="hw_makpc"
+  lircd_conf="makpc/lircd.conf.makpc"
+  lircmd_conf="makpc/lircmd.conf.makpc"
+fi
 if test "$driver" = "caraca"; then
   lirc_driver="caraca"
   hw_module="hw_caraca.o"
diff -ruN lirc-0.6.6.old/daemons/Makefile.am lirc-0.6.6/daemons/Makefile.am
--- lirc-0.6.6.old/daemons/Makefile.am	Mon Apr 22 21:07:47 2002
+++ lirc-0.6.6/daemons/Makefile.am	Sat Mar  1 12:09:57 2003
@@ -18,6 +18,7 @@
 			hw_pixelview.c hw_pixelview.h \
 			hw_pinsys.c hw_pinsys.h \
 			hw_logitech.c hw_logitech.h \
+			hw_makpc.c hw_makpc.h \
 			hw_slinke.c hw_slinke.h \
 			hw_irman.c hw_irman.h \
 			hw_creative.c hw_creative.h \
@@ -41,6 +42,7 @@
 			hw_pixelview.c hw_pixelview.h \
 			hw_pinsys.c hw_pinsys.h \
 			hw_logitech.c hw_logitech.h \
+			hw_makpc.c hw_makpc.h \
 			hw_slinke.c hw_slinke.h \
 			hw_irman.c hw_irman.h \
 			hw_creative.c hw_creative.h \
diff -ruN lirc-0.6.6.old/daemons/Makefile.in lirc-0.6.6/daemons/Makefile.in
--- lirc-0.6.6.old/daemons/Makefile.in	Sun Oct  6 11:24:53 2002
+++ lirc-0.6.6/daemons/Makefile.in	Sat Mar  1 12:12:23 2003
@@ -117,6 +117,7 @@
 			hw_pixelview.c hw_pixelview.h \
 			hw_pinsys.c hw_pinsys.h \
 			hw_logitech.c hw_logitech.h \
+			hw_makpc.c hw_makpc.h \
 			hw_slinke.c hw_slinke.h \
 			hw_irman.c hw_irman.h \
 			hw_creative.c hw_creative.h \
@@ -142,6 +143,7 @@
 			hw_pixelview.c hw_pixelview.h \
 			hw_pinsys.c hw_pinsys.h \
 			hw_logitech.c hw_logitech.h \
+			hw_makpc.c hw_makpc.h \
 			hw_slinke.c hw_slinke.h \
 			hw_irman.c hw_irman.h \
 			hw_creative.c hw_creative.h \
@@ -257,7 +259,7 @@
 @AMDEP_TRUE@	$(DEPDIR)/lircd_simsend-transmit.Po \
 @AMDEP_TRUE@	$(DEPDIR)/lircmd.Po $(DEPDIR)/receive.Po \
 @AMDEP_TRUE@	$(DEPDIR)/serial.Po $(DEPDIR)/slinke.Po \
-@AMDEP_TRUE@	$(DEPDIR)/transmit.Po
+@AMDEP_TRUE@	$(DEPDIR)/transmit.Po $(DEPDIR)/hw_makpc.Po
 COMPILE = $(CC) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) $(AM_CPPFLAGS) \
 	$(CPPFLAGS) $(AM_CFLAGS) $(CFLAGS)
 LTCOMPILE = $(LIBTOOL) --mode=compile $(CC) $(DEFS) $(DEFAULT_INCLUDES) \
diff -ruN lirc-0.6.6.old/daemons/hw-types.c lirc-0.6.6/daemons/hw-types.c
--- lirc-0.6.6.old/daemons/hw-types.c	Sat Jul 27 09:58:40 2002
+++ lirc-0.6.6/daemons/hw-types.c	Sat Mar  1 13:11:50 2003
@@ -10,6 +10,7 @@
 extern struct hardware hw_creative;
 extern struct hardware hw_irman;
 extern struct hardware hw_logitech;
+extern struct hardware hw_makpc;
 extern struct hardware hw_pinsys;
 extern struct hardware hw_pixelview;
 extern struct hardware hw_silitek;
@@ -49,6 +50,7 @@
 	&hw_irman,
 #endif
 	&hw_logitech,
+	&hw_makpc,
 	&hw_pinsys,
 	&hw_pixelview,
 	&hw_silitek,
diff -ruN lirc-0.6.6.old/daemons/hw_makpc.c lirc-0.6.6/daemons/hw_makpc.c
--- lirc-0.6.6.old/daemons/hw_makpc.c	Thu Jan  1 01:00:00 1970
+++ lirc-0.6.6/daemons/hw_makpc.c	Sat Mar  1 12:07:56 2003
@@ -0,0 +1,198 @@
+/*      $Id$      */
+
+/****************************************************************************
+ ** hw_makpc.c ***********************************************************
+ ****************************************************************************
+ *
+ * routines for Logitech receiver 
+ * 
+ * Copyright (C) 1999 Christoph Bartelmus <lirc@bartelmus.de>
+ * 	modified for logitech receiver by Isaac Lauer <inl101@alumni.psu.edu>
+ * 	modified for makpc receiver Pawel Ruchaj <p.ruchaj@elmak.pl>
+ *
+ */
+
+#ifdef HAVE_CONFIG_H
+# include <config.h>
+#endif
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <limits.h>
+#include <signal.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <sys/ioctl.h>
+
+#include "hardware.h"
+#include "serial.h"
+#include "ir_remote.h"
+#include "lircd.h"
+#include "hw_makpc.h"
+
+#define NUMBYTES 16 
+#define TIMEOUT 50000
+
+extern struct ir_remote *repeat_remote,*last_remote;
+
+unsigned char b[NUMBYTES];
+struct timeval start,end,last;
+lirc_t gap,signal_length;
+ir_code pre,code;
+int iNewKeyCode, repeat_counter,old_key;
+
+struct hardware hw_makpc=
+{
+	LIRC_DRIVER_DEVICE,       /* default device */
+	-1,                       /* fd */
+	LIRC_CAN_REC_LIRCCODE,    /* features */
+	0,                        /* send_mode */
+	LIRC_MODE_LIRCCODE,       /* rec_mode */
+	16,                       /* code_length */
+	makpc_init,            /* init_func */
+	NULL,                     /* config_func */
+	makpc_deinit,          /* deinit_func */
+	NULL,                     /* send_func */
+	makpc_rec,             /* rec_func */
+	makpc_decode,          /* decode_func */
+	NULL,                     /* readdata */
+	"makpc"
+};
+
+int makpc_decode(struct ir_remote *remote,
+		  ir_code *prep,ir_code *codep,ir_code *postp,
+		  int *repeat_flagp,lirc_t *remaining_gapp)
+{
+	if(!map_code(remote,prep,codep,postp,
+		     8,pre,8,code,0,0))
+	{
+		return(0);
+	}
+
+	gap=0;
+	if(start.tv_sec-last.tv_sec>=2) /* >1 sec */
+	{
+		*repeat_flagp=0;
+	}
+	else
+	{
+		gap=(start.tv_sec-last.tv_sec)*1000000+
+		start.tv_usec-last.tv_usec;
+		
+		if(gap<remote->remaining_gap*(100+remote->eps)/100
+		   || gap<=remote->remaining_gap+remote->aeps)
+			*repeat_flagp=1;
+		else
+			*repeat_flagp=0;
+	}
+	
+	*remaining_gapp=is_const(remote) ?
+	(remote->gap>signal_length ? remote->gap-signal_length:0):
+	remote->gap;
+
+	LOGPRINTF(1,"pre: %llx",(unsigned long long) *prep);
+	LOGPRINTF(1,"code: %llx",(unsigned long long) *codep);
+	LOGPRINTF(1,"repeat_flag: %d",*repeat_flagp);
+	LOGPRINTF(1,"gap: %lu",(unsigned long) gap);
+	LOGPRINTF(1,"rem: %lu",(unsigned long) remote->remaining_gap);
+	LOGPRINTF(1,"signal length: %lu",(unsigned long) signal_length);
+
+	return(1);
+}
+
+int makpc_init(void)
+{
+	signal_length=hw.code_length*1000000/1200;
+	
+	if(!tty_create_lock(hw.device))
+	{
+		logprintf(LOG_ERR,"could not create lock files");
+		return(0);
+	}
+	if((hw.fd=open(hw.device,O_RDWR|O_NONBLOCK|O_NOCTTY))<0)
+	{
+		logprintf(LOG_ERR,"could not open %s",hw.device);
+		logperror(LOG_ERR,"makpc_init()");
+		tty_delete_lock();
+		return(0);
+	}
+	if(!tty_reset(hw.fd))
+	{
+		logprintf(LOG_ERR,"could not reset tty");
+		makpc_deinit();
+		return(0);
+	}
+	if(!tty_setbaud(hw.fd,1200))
+	{
+		logprintf(LOG_ERR,"could not set baud rate");
+		makpc_deinit();
+		return(0);
+	}
+	return(1);
+}
+
+int makpc_deinit(void)
+{
+	close(hw.fd);
+	tty_delete_lock();
+	return(1);
+}
+
+char *makpc_rec(struct ir_remote *remotes)
+{
+	char *m;
+	int i=0;
+	int repeat, mouse_event;
+	int allow = 0;
+	b[i]=0x00;
+
+	last=end;
+	gettimeofday(&start,NULL);
+	if (b[i] == 0xAA) {iNewKeyCode = 1;repeat_counter = 0;}
+
+	while(b[i] != 0xAA)
+	{
+		i++;
+		if(i>=NUMBYTES) { LOGPRINTF(0,"buffer overflow",i); break;}
+		if(i>0) { if(!waitfordata(TIMEOUT)) { LOGPRINTF(0,"timeout reading byte %d",i); return(NULL); } }
+
+		if(read(hw.fd,&b[i],1)!=1)       { logprintf(LOG_ERR,"reading of byte %d failed",i);  logperror(LOG_ERR,NULL); return(NULL); }
+		if (b[i] == 0xAA) {iNewKeyCode = 1;repeat_counter = 0;}
+
+		LOGPRINTF(1,"byte %d: %02x",i,b[i]);
+		if((b[i] >= 0x01 && b[i] <= 0x2B) || (b[i] >= 0x41 && b[i] <= 0x6B)) //Range of the button codes
+		{
+			if(repeat_counter <1) {
+			  repeat_counter ++;
+				old_key = b[i];
+				allow = 0;
+			}
+			else {
+				if( old_key == b[i]) {
+					allow = 1;
+					mouse_event=b[i];
+					b[1]=0xA0;
+					b[2]=mouse_event;
+  					//printf("posz³o byte %d: %02x\n",i,b[i]);
+    					LOGPRINTF(1,"mouse event: %02x",mouse_event);
+				}
+				break;
+			}
+		}
+	}
+	if( allow) {
+		gettimeofday(&end,NULL);
+		if(b[1] == 0xA0) repeat=0;
+		else repeat=1;
+		if(!repeat) pre=(ir_code) b[1];
+		else pre=0xA0;
+		code=(ir_code) b[2];
+
+		m=decode_all(remotes);
+		return(m);
+	}
+	else
+	return 0;
+}
diff -ruN lirc-0.6.6.old/daemons/hw_makpc.h lirc-0.6.6/daemons/hw_makpc.h
--- lirc-0.6.6.old/daemons/hw_makpc.h	Thu Jan  1 01:00:00 1970
+++ lirc-0.6.6/daemons/hw_makpc.h	Sat Mar  1 12:08:48 2003
@@ -0,0 +1,26 @@
+/*      $Id$      */
+
+/****************************************************************************
+ ** hw_makpc.h **********************************************************
+ ****************************************************************************
+ *
+ * routines for Logitech receiver 
+ * 
+ * Copyright (C) 1999 Christoph Bartelmus <lirc@bartelmus.de>
+ *	modified for logitech receiver by Isaac Lauer <inl101@alumni.psu.edu>
+ *      modified for makpc receiver Pawel Ruchaj <p.ruchaj@elmak.pl>
+*/
+
+#ifndef _HW_MAKPC_H
+#define _HW_MAKPC_H
+
+#include "drivers/lirc.h"
+
+int makpc_decode(struct ir_remote *remote,
+		  ir_code *prep,ir_code *codep,ir_code *postp,
+		  int *repeat_flagp,lirc_t *remaining_gapp);
+int makpc_init(void);
+int makpc_deinit(void);
+char *makpc_rec(struct ir_remote *remotes);
+
+#endif
diff -ruN lirc-0.6.6.old/remotes/makpc/lircd.conf.makpc lirc-0.6.6/remotes/makpc/lircd.conf.makpc
--- lirc-0.6.6.old/remotes/makpc/lircd.conf.makpc	Thu Jan  1 01:00:00 1970
+++ lirc-0.6.6/remotes/makpc/lircd.conf.makpc	Mon Jan 20 10:06:12 2003
@@ -0,0 +1,114 @@
+#
+#
+# contributed by Pablo
+#
+# brand:             MAK PC
+# model:             Value Infrared Remote Control
+# supported devices: MAK PC receiver
+#
+
+begin remote
+
+  name  MAK_PC
+  bits           8
+  flags SPACE_ENC|CONST_LENGTH
+  eps           15
+  aeps          100
+
+  one             0   833
+  zero          833     0
+  pre_data_bits   8
+  pre_data       0xA0
+  gap           100000
+  repeat_bit      0
+
+      begin codes
+          1                        0x0000000000000001
+          2                        0x0000000000000002
+          3                        0x0000000000000003
+          4                        0x0000000000000004
+          5                        0x0000000000000005
+          6                        0x0000000000000006
+          7                        0x0000000000000007
+          8                        0x0000000000000008
+          9                        0x0000000000000009
+          0                        0x000000000000000A
+          TAB                      0x000000000000000B
+          ESC                      0x000000000000000C
+          ALT                      0x000000000000000D
+          PLAYLIST                 0x000000000000000E
+          ENTER                    0x000000000000000F
+          CH+                      0x0000000000000010
+          CH+VOLUP                 0x0000000000000011
+          VOLUP                    0x0000000000000012
+          CH-VOLUP                 0x0000000000000013
+          CH-                      0x0000000000000014
+          CH-VOLDOWN               0x0000000000000015
+          VOLDOWN                  0x0000000000000016
+          CH+VOLDOWN               0x0000000000000017
+          OK                       0x0000000000000018
+          BUTTONL_DOWN             0x0000000000000019
+          BUTTONR_DOWN             0x000000000000001A
+          POWER                    0x000000000000001B
+          A                        0x000000000000001C
+          B                        0x000000000000001D
+          C                        0x000000000000001E
+          D                        0x000000000000001F
+          BACK                     0x0000000000000020
+          PLAY                     0x0000000000000021
+          FORW                     0x0000000000000022
+          SKIP_BACK                0x0000000000000023
+          STOP                     0x0000000000000024
+          SKIP_FORW                0x0000000000000025
+          LOAD                     0x0000000000000026
+          PAUSE                    0x0000000000000027
+          FULLSCREEN               0x0000000000000028
+          MUTE                     0x0000000000000029
+          ALTVOL_UP                0x000000000000002A
+          ALTVOL_DOWN              0x000000000000002B
+          HOME                     0x0000000000000041
+          ARROW_UP                 0x0000000000000042
+          PG_UP                    0x0000000000000043
+          ARROW_LEFT               0x0000000000000044
+          WIN                      0x0000000000000045
+          ARROW_RIGHT              0x0000000000000046
+          END                      0x0000000000000047
+          ARROW_DOWN               0x0000000000000048
+          PG_DN                    0x0000000000000049
+          SPACE                    0x000000000000004A
+          sTAB                     0x000000000000004B
+          sESC                     0x000000000000004C
+          sALT                     0x000000000000004D
+          sPLAYLIST                0x000000000000004E
+          sENTER                   0x000000000000004F
+          MOUSE_N                  0x0000000000000050
+          MOUSE_NE                 0x0000000000000051
+          MOUSE_E                  0x0000000000000052
+          MOUSE_SE                 0x0000000000000053
+          MOUSE_S                  0x0000000000000054
+          MOUSE_SW                 0x0000000000000055
+          MOUSE_W                  0x0000000000000056
+          MOUSE_NW                 0x0000000000000057
+          sOK                      0x0000000000000058
+          BUTTONL_UP               0x0000000000000059
+          BUTTONR_UP               0x000000000000005A
+          EXIT                     0x000000000000005B
+          E                        0x000000000000005C
+          F                        0x000000000000005D
+          G                        0x000000000000005E
+          H                        0x000000000000005F
+          sREV                     0x0000000000000060
+          sPLAY                    0x0000000000000061
+          sFFD                     0x0000000000000062
+          sPREV                    0x0000000000000063
+          sSTOP                    0x0000000000000064
+          sNEXT                    0x0000000000000065
+          TXT                      0x0000000000000066
+          sPAUSE                   0x0000000000000067 
+          sFULLSCREEN              0x0000000000000068 
+          ZOOM                     0x0000000000000069 
+          ZOOM-                    0x000000000000006A 
+          ZOOM+                    0x000000000000006B 
+      end codes
+
+end remote
diff -ruN lirc-0.6.6.old/remotes/makpc/lircmd.conf.makpc lirc-0.6.6/remotes/makpc/lircmd.conf.makpc
--- lirc-0.6.6.old/remotes/makpc/lircmd.conf.makpc	Thu Jan  1 01:00:00 1970
+++ lirc-0.6.6/remotes/makpc/lircmd.conf.makpc	Mon Jan 20 11:27:21 2003
@@ -0,0 +1,29 @@
+# 
+# lircmd config file for the MAK_PC control
+# 
+
+PROTOCOL IntelliMouse
+
+# ACCELERATOR start max multiplier
+
+ACCELERATOR 1 25 2
+
+#ACTIVATE * MENU
+
+MOVE_N    MAK_PC   MOUSE_N
+MOVE_NE   MAK_PC   MOUSE_NE
+
+MOVE_E    MAK_PC   MOUSE_E
+MOVE_SE   MAK_PC   MOUSE_SE
+
+MOVE_S    MAK_PC   MOUSE_S
+MOVE_SW   MAK_PC   MOUSE_SW
+
+MOVE_W    MAK_PC   MOUSE_W
+MOVE_NW   MAK_PC   MOUSE_NW
+
+BUTTON1_DOWN   MAK_PC   BUTTONL_DOWN
+BUTTON3_DOWN   MAK_PC   BUTTONR_DOWN
+
+BUTTON1_UP   MAK_PC   BUTTONL_UP
+BUTTON3_UP   MAK_PC   BUTTONR_UP
diff -ruN lirc-0.6.6.old/setup.sh lirc-0.6.6/setup.sh
--- lirc-0.6.6.old/setup.sh	Sun Oct  6 10:56:29 2002
+++ lirc-0.6.6/setup.sh	Sat Mar  1 12:24:10 2003
@@ -160,7 +160,8 @@
     elif test "$LIRC_DRIVER" = "remotemaster" -o "$LIRC_DRIVER" = "irman" -o \
 	"$LIRC_DRIVER" = "logitech" -o "$LIRC_DRIVER" = "pctv" -o \
 	"$LIRC_DRIVER" = "creative" -o "$LIRC_DRIVER" = "slinke" -o \
-        "$LIRC_DRIVER" = "silitek" -o "$LIRC_DRIVER" = "realmagic"; then
+        "$LIRC_DRIVER" = "silitek" -o "$LIRC_DRIVER" = "realmagic" -o \
+	"$LIRC_DRIVER" = "makpc"; then
 	{
         dialog --clear --backtitle "$BACKTITLE" \
                --title "Select tty to use" \
@@ -277,7 +278,8 @@
 			b "REALmagic (bundled with Hollywood Plus DVD card)" \
 			c "Slink-e (receive only)" \
                         d "Silitek SM-1000" \
-			e "Tekram Irmate 210 (16x50 UART compatible serial port)" 2> $TEMP;
+			e "Tekram Irmate 210 (16x50 UART compatible serial port)" \
+			f "MakPC" 2> $TEMP;
 	    if test "$?" = "0"; then
 		{
 		set `cat $TEMP`
@@ -296,6 +298,7 @@
 		elif test "$1" = "c"; then LIRC_DRIVER=slinke;       DRIVER_PARAMETER=tty3;
 		elif test "$1" = "d"; then LIRC_DRIVER=silitek;      DRIVER_PARAMETER=tty1;
 		elif test "$1" = "e"; then LIRC_DRIVER=tekram;       DRIVER_PARAMETER=com1;
+		elif test "$1" = "f"; then LIRC_DRIVER=makpc;        DRIVER_PARAMETER=tty1;
 		fi
 		}
 	    else
@@ -492,7 +495,7 @@
     elif test "$LIRC_DRIVER" = "irman" -o "$LIRC_DRIVER" = "remotemaster" -o \
 	"$LIRC_DRIVER" = "logitech" -o "$LIRC_DRIVER" = "pctv" -o \
 	"$LIRC_DRIVER" = "slinke" -o "$LIRC_DRIVER" = "silitek" -o \
-        "$LIRC_DRIVER" = "realmagic"; then
+        "$LIRC_DRIVER" = "realmagic" -o "$LIRC_DRIVER" = "makpc"; then
         {
 	echo "--with-tty=$IRTTY \\" >>$START
 	}
