diff -urN lirc-0.6.3.org/drivers/lirc_parallel/lirc_parallel.c lirc-0.6.3/drivers/lirc_parallel/lirc_parallel.c
--- lirc-0.6.3.org/drivers/lirc_parallel/lirc_parallel.c	Thu Jul 12 00:41:24 2001
+++ lirc-0.6.3/drivers/lirc_parallel/lirc_parallel.c	Fri Jul 13 10:15:55 2001
@@ -45,6 +45,7 @@
 #include <asm/io.h>
 #include <asm/signal.h>
 #include <asm/irq.h>
+#include <asm/spinlock.h>
 
 #ifdef KERNEL_2_2
 #include <asm/uaccess.h>
@@ -447,6 +448,7 @@
 	unsigned int i;
 	unsigned int level,newlevel;
 	unsigned long flags;
+	spinlock_t irq_lock = SPIN_LOCK_UNLOCKED;
 	lirc_t counttimer;
 	
 #ifdef KERNEL_2_2
@@ -486,7 +488,7 @@
 		wbuf[i]=(lirc_t) (((double) wbuf[i])*timer/1000000);
 	}
 	
-	save_flags(flags);cli();
+	spin_lock_irqsave(&irq_lock, flags);
 	i=0;
 	while(i<count)
 	{
@@ -501,7 +503,7 @@
 			if(in(1)&LP_PSELECD)
 			{
 				lirc_off();
-				restore_flags(flags); /* sti(); */
+				spin_unlock_irqrestore(&irq_lock, flags); /* sti(); */
 				return(-EIO);
 			}
 		}
@@ -517,13 +519,13 @@
 			level=newlevel;
 			if(in(1)&LP_PSELECD)
 			{
-				restore_flags(flags); /* sti(); */
+				spin_unlock_irqrestore(&irq_lock, flags); /* sti(); */
 				return(-EIO);
 			}
 		}
 		while(counttimer<wbuf[i]);i++;
 	}
-	restore_flags(flags); /* sti(); */
+	spin_unlock_irqrestore(&irq_lock, flags); /* sti(); */
 #else
 	/* 
 	   place code that handles write
diff -urN lirc-0.6.3.org/drivers/lirc_serial/lirc_serial.c lirc-0.6.3/drivers/lirc_serial/lirc_serial.c
--- lirc-0.6.3.org/drivers/lirc_serial/lirc_serial.c	Thu Jul 12 00:41:24 2001
+++ lirc-0.6.3/drivers/lirc_serial/lirc_serial.c	Fri Jul 13 10:12:32 2001
@@ -66,6 +66,7 @@
 #include <asm/segment.h>
 #include <asm/io.h>
 #include <asm/irq.h>
+#include <asm/spinlock.h>
 #include <asm/fcntl.h>
 
 #include "drivers/lirc.h"
@@ -439,6 +440,7 @@
 static int init_port(void)
 {
 	unsigned long flags;
+	spinlock_t irq_lock = SPIN_LOCK_UNLOCKED;
 
         /* Check io region*/
 	
@@ -464,7 +466,7 @@
 	/* Reserve io region. */
 	request_region(port, 8, LIRC_DRIVER_NAME);
 	
-	save_flags(flags);cli();
+	spin_lock_irqsave(&irq_lock, flags);
 	
 	/* Set DLAB 0. */
 	soutp(UART_LCR, sinp(UART_LCR) & (~UART_LCR_DLAB));
@@ -502,7 +504,7 @@
 	/* THR interrupt already disabled at this point */
 #endif
 	
-	restore_flags(flags);
+	spin_unlock_irqrestore(&irq_lock, flags);
 	
 	/* If pin is high, then this must be an active low receiver. */
 	if(sense==-1)
@@ -538,6 +540,7 @@
 {
 	int result;
 	unsigned long flags;
+	spinlock_t irq_lock = SPIN_LOCK_UNLOCKED;
 	
 #       ifdef KERNEL_2_1
 	spin_lock(&lirc_lock);
@@ -578,14 +581,14 @@
 	};
 
 	/* finally enable interrupts. */
-	save_flags(flags);cli();
+	spin_lock_irqsave(&irq_lock, flags);
 	
 	/* Set DLAB 0. */
 	soutp(UART_LCR, sinp(UART_LCR) & (~UART_LCR_DLAB));
 	
 	soutp(UART_IER, sinp(UART_IER)|UART_IER_MSI);
 	
-	restore_flags(flags);
+	spin_unlock_irqrestore(&irq_lock, flags);
 	
 	/* Init read buffer pointers. */
 	rbh = rbt = 0;
@@ -603,8 +606,9 @@
 static void lirc_close(struct inode *node, struct file *file)
 #endif
 {	unsigned long flags;
+    	spinlock_t irq_lock = SPIN_LOCK_UNLOCKED;
 	
-	save_flags(flags);cli();
+	spin_lock_irqsave(&irq_lock, flags);
 	
 	/* Set DLAB 0. */
 	soutp(UART_LCR, sinp(UART_LCR) & (~UART_LCR_DLAB));
@@ -612,7 +616,7 @@
 	/* First of all, disable all interrupts */
 	soutp(UART_IER, sinp(UART_IER)&
 	      (~(UART_IER_MSI|UART_IER_RLSI|UART_IER_THRI|UART_IER_RDI)));
-	restore_flags(flags);
+	spin_unlock_irqrestore(&irq_lock, flags);
 	
 	free_irq(irq, NULL);
 #       ifdef DEBUG
@@ -712,6 +716,7 @@
 #ifdef LIRC_SERIAL_TRANSMITTER
 	int retval,i,count;
 	unsigned long flags;
+	spinlock_t irq_lock = SPIN_LOCK_UNLOCKED;
 	
 	if(n%sizeof(lirc_t)) return(-EINVAL);
 	retval=verify_area(VERIFY_READ,buf,n);
@@ -723,7 +728,7 @@
 #       else
 	memcpy_fromfs(wbuf,buf,n);
 #       endif
-	save_flags(flags);cli();
+	spin_lock_irqsave(&irq_lock, flags);
 #       ifdef LIRC_SERIAL_IRDEO
 	/* DTR, RTS down */
 	on();
@@ -734,7 +739,7 @@
 		else send_pulse(wbuf[i]);
 	}
 	off();
-	restore_flags(flags);
+	spin_unlock_irqrestore(&irq_lock, flags);
 	return(n);
 #else
 	return(-EBADF);
