diff -Nur lirc-0.8.0.org/drivers/lirc_dev/lirc_dev.c lirc-0.8.0/drivers/lirc_dev/lirc_dev.c
--- lirc-0.8.0.org/drivers/lirc_dev/lirc_dev.c	2006-06-03 00:51:36.000000000 +0200
+++ lirc-0.8.0/drivers/lirc_dev/lirc_dev.c	2006-06-03 00:52:01.000000000 +0200
@@ -469,7 +469,7 @@
 		{
 			struct task_struct *p;
 			
-			p = find_task_by_pid(ir->tpid);
+			p = find_task_by_real_pid(ir->tpid);
 			wake_up_process(p);
 		}
 #else
