--- lirc-0.6.5/drivers/lirc_gpio/lirc_gpio.c.orig	Wed Dec 12 23:27:38 2001
+++ lirc-0.6.5/drivers/lirc_gpio/lirc_gpio.c	Thu Aug 29 16:43:42 2002
@@ -33,6 +33,10 @@
 #error "*******************************************************"
 #endif
 
+#ifdef __alpha__
+#include <asm/current.h>
+#endif
+
 #include <linux/module.h>
 #include <linux/kmod.h>
 #include <linux/sched.h>
--- lirc-0.6.5/drivers/lirc_i2c/lirc_i2c.c.orig	Wed Dec 12 23:27:39 2001
+++ lirc-0.6.5/drivers/lirc_i2c/lirc_i2c.c	Thu Aug 29 16:47:53 2002
@@ -37,6 +37,10 @@
 #include <config.h>
 #endif
 
+#ifdef __alpha__
+#include <asm/current.h>
+#endif
+
 #include <linux/module.h>
 #include <linux/kmod.h>
 #include <linux/kernel.h>
--- lirc-0.6.5/drivers/lirc_serial/lirc_serial.c.orig	Thu Aug 29 16:48:17 2002
+++ lirc-0.6.5/drivers/lirc_serial/lirc_serial.c	Thu Aug 29 16:51:36 2002
@@ -66,6 +66,10 @@
 #define LIRC_LOOPS_PER_JIFFY
 #endif
 
+#ifdef __alpha__
+#include <asm/current.h>
+#endif
+
 #include <linux/config.h>
 
 #ifndef CONFIG_SERIAL_MODULE
--- lirc-0.6.5/drivers/lirc_sir/lirc_sir.c.orig	Thu Aug 29 16:51:58 2002
+++ lirc-0.6.5/drivers/lirc_sir/lirc_sir.c	Thu Aug 29 16:53:58 2002
@@ -44,6 +44,10 @@
 #define KERNEL_2_0
 #endif
 
+#ifdef __alpha__
+#include <asm/current.h>
+#endif
+
 #include <linux/module.h>
 
 #ifdef HAVE_CONFIG_H
--- lirc-0.6.5/drivers/lirc_parallel/lirc_parallel.c.orig	Thu Aug 29 16:54:19 2002
+++ lirc-0.6.5/drivers/lirc_parallel/lirc_parallel.c	Thu Aug 29 16:58:31 2002
@@ -48,6 +48,10 @@
 #error "--- Sorry, this driver is not SMP safe. ---"
 #endif
 
+#ifdef __alpha__
+#include <asm/current.h>
+#endif
+
 #include <linux/module.h>
 #include <linux/sched.h>
 #include <linux/errno.h>
--- lirc-0.6.5/drivers/lirc_dev/lirc_dev.c.orig	Thu Aug 29 17:00:01 2002
+++ lirc-0.6.5/drivers/lirc_dev/lirc_dev.c	Thu Aug 29 17:02:00 2002
@@ -36,6 +36,10 @@
 #error "********************************************************"
 #endif
 
+#ifdef __alpha__
+#include <asm/current.h>
+#endif
+
 #include <linux/config.h>
 #include <linux/module.h>
 #include <linux/kernel.h>
