diff -urN lirc-0.7.1pre1.orig/drivers/lirc_i2c/lirc_i2c.c lirc-0.7.1pre1/drivers/lirc_i2c/lirc_i2c.c
--- lirc-0.7.1pre1.orig/drivers/lirc_i2c/lirc_i2c.c	2005-01-14 17:07:12.000000000 +0100
+++ lirc-0.7.1pre1/drivers/lirc_i2c/lirc_i2c.c	2005-01-25 11:07:08.527467616 +0100
@@ -268,7 +268,7 @@
 	/* lock bttv in memory while /dev/lirc is in use  */
 	/* this is completely broken code. lirc_unregister_plugin()
 	   must be possible even when the device is open */
-	i2c_use_client(&ir->c);
+//	i2c_use_client(&ir->c);
 
 	MOD_INC_USE_COUNT;
 	return 0;
@@ -276,9 +276,9 @@
 
 static void set_use_dec(void* data)
 {
-	struct IR *ir = data;
+//	struct IR *ir = data;
 
-	i2c_release_client(&ir->c);
+//	i2c_release_client(&ir->c);
 	MOD_DEC_USE_COUNT;
 }
 
@@ -384,7 +384,7 @@
 	/* register device */
 	i2c_attach_client(&ir->c);
 	ir->l.minor = lirc_register_plugin(&ir->l);
-	i2c_use_client(&ir->c);
+//	i2c_use_client(&ir->c);
 	
 	return 0;
 }
@@ -394,7 +394,7 @@
 	struct IR *ir = i2c_get_clientdata(client);
 	
 	/* unregister device */
-	i2c_release_client(&ir->c);
+//	i2c_release_client(&ir->c);
 	lirc_unregister_plugin(ir->l.minor);
 	i2c_detach_client(&ir->c);
 
