--- lirc-0.8.3/drivers/lirc_parallel/lirc_parallel.c.orig	2008-05-05 19:17:41.000000000 +0200
+++ lirc-0.8.3/drivers/lirc_parallel/lirc_parallel.c	2008-05-05 19:17:28.000000000 +0200
@@ -71,6 +71,14 @@
 #include <linux/io.h>
 #include <linux/signal.h>
 #include <linux/irq.h>
+
+#if defined(__sparc__) && defined(enable_irq)
+#  undef enable_irq
+#endif
+#if defined(__sparc__) && defined(disable_irq)
+#  undef disable_irq
+#endif
+
 #include <linux/uaccess.h>
 #endif
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 4, 0)
