diff -urN lirc-0.7.1pre1.orig/drivers/lirc_atiusb/lirc_atiusb.c lirc-0.7.1pre1/drivers/lirc_atiusb/lirc_atiusb.c
--- lirc-0.7.1pre1.orig/drivers/lirc_atiusb/lirc_atiusb.c	2004-12-31 13:58:16.000000000 +0100
+++ lirc-0.7.1pre1/drivers/lirc_atiusb/lirc_atiusb.c	2005-01-25 10:27:53.768445400 +0100
@@ -39,6 +39,10 @@
 #error "*******************************************************"
 #endif
 
+#ifdef __alpha__
+#include <asm/current.h>
+#endif
+
 #include <linux/autoconf.h>
 
 #include <linux/kernel.h>
diff -urN lirc-0.7.1pre1.orig/drivers/lirc_bt829/lirc_bt829.c lirc-0.7.1pre1/drivers/lirc_bt829/lirc_bt829.c
--- lirc-0.7.1pre1.orig/drivers/lirc_bt829/lirc_bt829.c	2004-08-07 12:06:07.000000000 +0200
+++ lirc-0.7.1pre1/drivers/lirc_bt829/lirc_bt829.c	2005-01-25 10:27:53.770445096 +0100
@@ -23,6 +23,10 @@
 #error "This driver needs kernel version 2.4.0 or higher"
 #endif
 
+#ifdef __alpha__
+#include <asm/current.h>
+#endif
+
 #include <linux/autoconf.h>
 #include <linux/kernel.h>
 #include <linux/module.h>
diff -urN lirc-0.7.1pre1.orig/drivers/lirc_dev/lirc_dev.c lirc-0.7.1pre1/drivers/lirc_dev/lirc_dev.c
--- lirc-0.7.1pre1.orig/drivers/lirc_dev/lirc_dev.c	2004-09-05 18:48:48.000000000 +0200
+++ lirc-0.7.1pre1/drivers/lirc_dev/lirc_dev.c	2005-01-25 10:27:53.771444944 +0100
@@ -32,6 +32,10 @@
 #error "**********************************************************"
 #endif
 
+#ifdef __alpha__
+#include <asm/current.h>
+#endif
+
 #include <linux/autoconf.h>
 #include <linux/module.h>
 #include <linux/kernel.h>
diff -urN lirc-0.7.1pre1.orig/drivers/lirc_gpio/lirc_gpio.c lirc-0.7.1pre1/drivers/lirc_gpio/lirc_gpio.c
--- lirc-0.7.1pre1.orig/drivers/lirc_gpio/lirc_gpio.c	2004-08-07 12:06:08.000000000 +0200
+++ lirc-0.7.1pre1/drivers/lirc_gpio/lirc_gpio.c	2005-01-25 10:27:53.772444792 +0100
@@ -36,6 +36,10 @@
 #error "*******************************************************"
 #endif
 
+#ifdef __alpha__
+#include <asm/current.h>
+#endif
+
 #include <linux/version.h>
 #include <linux/module.h>
 #include <linux/kmod.h>
diff -urN lirc-0.7.1pre1.orig/drivers/lirc_i2c/lirc_i2c.c lirc-0.7.1pre1/drivers/lirc_i2c/lirc_i2c.c
--- lirc-0.7.1pre1.orig/drivers/lirc_i2c/lirc_i2c.c	2005-01-14 17:07:12.000000000 +0100
+++ lirc-0.7.1pre1/drivers/lirc_i2c/lirc_i2c.c	2005-01-25 10:31:17.166524192 +0100
@@ -37,6 +37,10 @@
 #include <config.h>
 #endif
 
+#ifdef __alpha__
+#include <asm/current.h>
+#endif
+
 #include <linux/version.h>
 #include <linux/module.h>
 #include <linux/kmod.h>
diff -urN lirc-0.7.1pre1.orig/drivers/lirc_igorplugusb/lirc_igorplugusb.c lirc-0.7.1pre1/drivers/lirc_igorplugusb/lirc_igorplugusb.c
--- lirc-0.7.1pre1.orig/drivers/lirc_igorplugusb/lirc_igorplugusb.c	2004-08-07 11:52:46.000000000 +0200
+++ lirc-0.7.1pre1/drivers/lirc_igorplugusb/lirc_igorplugusb.c	2005-01-25 10:27:53.777444032 +0100
@@ -47,6 +47,10 @@
 #error "*******************************************************"
 #endif
 
+#ifdef __alpha__
+#include <asm/current.h>
+#endif
+
 #include <linux/autoconf.h>
 #include <linux/module.h>
 #include <linux/kernel.h>
diff -urN lirc-0.7.1pre1.orig/drivers/lirc_it87/lirc_it87.c lirc-0.7.1pre1/drivers/lirc_it87/lirc_it87.c
--- lirc-0.7.1pre1.orig/drivers/lirc_it87/lirc_it87.c	2004-12-30 22:47:19.000000000 +0100
+++ lirc-0.7.1pre1/drivers/lirc_it87/lirc_it87.c	2005-01-25 10:27:53.779443728 +0100
@@ -31,6 +31,10 @@
 #include <linux/version.h>
 #include <linux/module.h>
 
+#ifdef __alpha__
+#include <asm/current.h>
+#endif
+
 #ifdef HAVE_CONFIG_H
 # include <config.h>
 #endif
diff -urN lirc-0.7.1pre1.orig/drivers/lirc_mceusb/lirc_mceusb.c lirc-0.7.1pre1/drivers/lirc_mceusb/lirc_mceusb.c
--- lirc-0.7.1pre1.orig/drivers/lirc_mceusb/lirc_mceusb.c	2004-09-05 18:48:49.000000000 +0200
+++ lirc-0.7.1pre1/drivers/lirc_mceusb/lirc_mceusb.c	2005-01-25 10:27:53.780443576 +0100
@@ -108,6 +108,10 @@
 #include <linux/signal.h>
 #endif
 
+#ifdef __alpha__
+#include <asm/current.h>
+#endif
+
 #ifdef CONFIG_USB_DEBUG
 	static int debug = 1;
 #else
diff -urN lirc-0.7.1pre1.orig/drivers/lirc_parallel/lirc_parallel.c lirc-0.7.1pre1/drivers/lirc_parallel/lirc_parallel.c
--- lirc-0.7.1pre1.orig/drivers/lirc_parallel/lirc_parallel.c	2005-01-22 11:42:39.000000000 +0100
+++ lirc-0.7.1pre1/drivers/lirc_parallel/lirc_parallel.c	2005-01-25 10:27:53.782443272 +0100
@@ -40,6 +40,10 @@
 #error "**********************************************************"
 #endif
 
+#ifdef __alpha__
+#include <asm/current.h>
+#endif
+
 #include <linux/autoconf.h>
 
 #ifdef CONFIG_SMP
diff -urN lirc-0.7.1pre1.orig/drivers/lirc_sasem/lirc_sasem.c lirc-0.7.1pre1/drivers/lirc_sasem/lirc_sasem.c
--- lirc-0.7.1pre1.orig/drivers/lirc_sasem/lirc_sasem.c	2004-11-07 14:35:45.000000000 +0100
+++ lirc-0.7.1pre1/drivers/lirc_sasem/lirc_sasem.c	2005-01-25 10:27:53.783443120 +0100
@@ -60,6 +60,10 @@
 
 #include <linux/autoconf.h>
 
+#ifdef __alpha__
+#include <asm/current.h>
+#endif
+
 #include <linux/errno.h>
 #include <linux/init.h>
 #include <linux/kernel.h>
diff -urN lirc-0.7.1pre1.orig/drivers/lirc_serial/lirc_serial.c lirc-0.7.1pre1/drivers/lirc_serial/lirc_serial.c
--- lirc-0.7.1pre1.orig/drivers/lirc_serial/lirc_serial.c	2004-12-31 22:21:39.000000000 +0100
+++ lirc-0.7.1pre1/drivers/lirc_serial/lirc_serial.c	2005-01-25 10:32:26.256020992 +0100
@@ -61,6 +61,10 @@
 #error "**********************************************************"
 #endif
 
+#ifdef __alpha__
+#include <asm/current.h>
+#endif
+
 #include <linux/autoconf.h>
 
 #if defined(CONFIG_SERIAL) || defined(CONFIG_SERIAL_8250)
diff -urN lirc-0.7.1pre1.orig/drivers/lirc_sir/lirc_sir.c lirc-0.7.1pre1/drivers/lirc_sir/lirc_sir.c
--- lirc-0.7.1pre1.orig/drivers/lirc_sir/lirc_sir.c	2004-12-30 22:38:16.000000000 +0100
+++ lirc-0.7.1pre1/drivers/lirc_sir/lirc_sir.c	2005-01-25 10:27:53.787442512 +0100
@@ -40,6 +40,11 @@
 #error " Sorry, this driver needs kernel version 2.2.18 or higher "
 #error "**********************************************************"
 #endif
+
+#ifdef __alpha__
+#include <asm/current.h>
+#endif
+
 #include <linux/module.h>
 
 #ifdef HAVE_CONFIG_H
