--- lirc-0.8.3/configure.ac.orig	2008-05-15 11:40:28.000000000 +0200
+++ lirc-0.8.3/configure.ac	2008-05-15 11:42:34.282299065 +0200
@@ -49,17 +49,6 @@
 fi
 
 vga_progs=
-AC_CHECK_HEADER(vga.h,
-  [AC_CHECK_HEADER(vgagl.h,
-    [AC_CHECK_LIB(vga, vga_init,
-      [AC_CHECK_LIB(vgagl, gl_setcontextvga,
-        [AC_DEFINE(HAVE_LIBVGA)]
-        vga_progs=smode2,,
-        -lvga
-      )]
-    )]
-  )]
-)
 x_progs="irxevent xmode2"
 AC_PATH_XTRA
 if test "$no_x" = "yes"; then
