diff -urN lirc-0.7.0.orig/drivers/lirc_parallel/lirc_parallel.c lirc-0.7.0/drivers/lirc_parallel/lirc_parallel.c
--- lirc-0.7.0.orig/drivers/lirc_parallel/lirc_parallel.c	2004-11-07 13:31:18.000000000 +0000
+++ lirc-0.7.0/drivers/lirc_parallel/lirc_parallel.c	2004-11-09 22:41:47.630409160 +0000
@@ -61,6 +61,14 @@
 #include <asm/io.h>
 #include <asm/signal.h>
 #include <asm/irq.h>
+
+#if defined(__sparc__) && defined(enable_irq)
+#  undef enable_irq
+#endif
+#if defined(__sparc__) && defined(disable_irq)
+#  undef disable_irq
+#endif
+
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 4, 0)
 #include <asm/div64.h>
 #endif
