--- lirc-0.9.0/./drivers/lirc_atiusb/lirc_atiusb.c~	2011-03-25 23:28:18.000000000 +0100
+++ lirc-0.9.0/./drivers/lirc_atiusb/lirc_atiusb.c	2011-09-10 11:19:50.422542829 +0200
@@ -48,7 +48,6 @@
 #include <linux/slab.h>
 #include <linux/module.h>
 #include <linux/kmod.h>
-#include <linux/smp_lock.h>
 #include <linux/completion.h>
 #include <linux/uaccess.h>
 #include <linux/usb.h>
--- lirc-0.9.0/drivers/lirc_i2c/lirc_i2c.c~	2011-03-25 23:28:18.000000000 +0100
+++ lirc-0.9.0/drivers/lirc_i2c/lirc_i2c.c	2011-09-10 12:24:55.710859996 +0200
@@ -555,8 +555,8 @@
 		kfree(ir);
 		return -EINVAL;
 	}
-	printk(KERN_INFO "lirc_i2c: chip 0x%x found @ 0x%02x (%s)\n",
-	       adap->id, addr, ir->c.name);
+	printk(KERN_INFO "lirc_i2c: chip found @ 0x%02x (%s)\n",
+	       addr, ir->c.name);
 
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 31)
 	/* register device */
--- lirc-0.9.0/drivers/lirc_gpio/lirc_gpio.c.orig	2011-03-25 23:28:18.000000000 +0100
+++ lirc-0.9.0/drivers/lirc_gpio/lirc_gpio.c	2011-09-10 16:30:47.051485373 +0200
@@ -38,7 +38,6 @@
 #include <linux/errno.h>
 
 #include "../drivers/media/video/bt8xx/bttv.h"
-#include "../drivers/media/video/bt8xx/bttvp.h"
 
 #include "drivers/kcompat.h"
 #if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 35)
