diff -aur lirc-0.8.7.orig/drivers/lirc_atiusb/lirc_atiusb.c lirc-0.8.7/drivers/lirc_atiusb/lirc_atiusb.c
--- lirc-0.8.7.orig/drivers/lirc_atiusb/lirc_atiusb.c	2010-07-25 18:43:33.000000000 +0200
+++ lirc-0.8.7/drivers/lirc_atiusb/lirc_atiusb.c	2010-09-21 13:05:27.000000000 +0200
@@ -43,6 +43,10 @@
 #error "*******************************************************"
 #endif
 
+#ifdef __alpha__
+#include <asm/current.h>
+#endif
+
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
 #include <linux/autoconf.h>
 #endif
diff -aur lirc-0.8.7.orig/drivers/lirc_bt829/lirc_bt829.c lirc-0.8.7/drivers/lirc_bt829/lirc_bt829.c
--- lirc-0.8.7.orig/drivers/lirc_bt829/lirc_bt829.c	2010-03-17 15:16:15.000000000 +0100
+++ lirc-0.8.7/drivers/lirc_bt829/lirc_bt829.c	2010-09-21 13:03:34.000000000 +0200
@@ -22,6 +22,11 @@
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 4, 0)
 #error "This driver needs kernel version 2.4.0 or higher"
 #endif
+
+#ifdef __alpha__
+#include <asm/current.h>
+#endif
+
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
 #include <linux/autoconf.h>
 #endif
diff -aur lirc-0.8.7.orig/drivers/lirc_dev/lirc_dev.c lirc-0.8.7/drivers/lirc_dev/lirc_dev.c
--- lirc-0.8.7.orig/drivers/lirc_dev/lirc_dev.c	2010-08-17 23:03:23.000000000 +0200
+++ lirc-0.8.7/drivers/lirc_dev/lirc_dev.c	2010-09-21 12:58:34.000000000 +0200
@@ -32,6 +32,10 @@
 #error "**********************************************************"
 #endif
 
+#ifdef __alpha__
+#include <asm/current.h>
+#endif
+
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
 #include <linux/autoconf.h>
 #endif
diff -aur lirc-0.8.7.orig/drivers/lirc_ene0100/lirc_ene0100.c lirc-0.8.7/drivers/lirc_ene0100/lirc_ene0100.c
--- lirc-0.8.7.orig/drivers/lirc_ene0100/lirc_ene0100.c	2010-08-16 22:20:46.000000000 +0200
+++ lirc-0.8.7/drivers/lirc_ene0100/lirc_ene0100.c	2010-09-21 12:53:12.000000000 +0200
@@ -19,6 +19,10 @@
  * USA
  */
 
+#ifdef __alpha__
+#include <asm/current.h>
+#endif
+
 #include <linux/kernel.h>
 #include <linux/module.h>
 #include <linux/pnp.h>
diff -aur lirc-0.8.7.orig/drivers/lirc_gpio/lirc_gpio.c lirc-0.8.7/drivers/lirc_gpio/lirc_gpio.c
--- lirc-0.8.7.orig/drivers/lirc_gpio/lirc_gpio.c	2010-04-26 21:47:37.000000000 +0200
+++ lirc-0.8.7/drivers/lirc_gpio/lirc_gpio.c	2010-09-21 12:53:12.000000000 +0200
@@ -36,6 +36,10 @@
 #error "*******************************************************"
 #endif
 
+#ifdef __alpha__
+#include <asm/current.h>
+#endif
+
 #include <linux/version.h>
 #include <linux/module.h>
 #include <linux/kmod.h>
diff -aur lirc-0.8.7.orig/drivers/lirc_i2c/lirc_i2c.c lirc-0.8.7/drivers/lirc_i2c/lirc_i2c.c
--- lirc-0.8.7.orig/drivers/lirc_i2c/lirc_i2c.c	2009-12-28 16:29:03.000000000 +0100
+++ lirc-0.8.7/drivers/lirc_i2c/lirc_i2c.c	2010-09-21 12:53:12.000000000 +0200
@@ -48,6 +48,10 @@
 #include <config.h>
 #endif
 
+#ifdef __alpha__
+#include <asm/current.h>
+#endif
+
 #include <linux/version.h>
 #include <linux/module.h>
 #include <linux/kmod.h>
diff -aur lirc-0.8.7.orig/drivers/lirc_igorplugusb/lirc_igorplugusb.c lirc-0.8.7/drivers/lirc_igorplugusb/lirc_igorplugusb.c
--- lirc-0.8.7.orig/drivers/lirc_igorplugusb/lirc_igorplugusb.c	2010-07-25 18:43:33.000000000 +0200
+++ lirc-0.8.7/drivers/lirc_igorplugusb/lirc_igorplugusb.c	2010-09-21 13:04:23.000000000 +0200
@@ -47,6 +47,10 @@
 #error "*******************************************************"
 #endif
 
+#ifdef __alpha__
+#include <asm/current.h>
+#endif
+
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
 #include <linux/autoconf.h>
 #endif
diff -aur lirc-0.8.7.orig/drivers/lirc_imon/lirc_imon.c lirc-0.8.7/drivers/lirc_imon/lirc_imon.c
--- lirc-0.8.7.orig/drivers/lirc_imon/lirc_imon.c	2010-07-11 04:59:43.000000000 +0200
+++ lirc-0.8.7/drivers/lirc_imon/lirc_imon.c	2010-09-21 13:04:42.000000000 +0200
@@ -30,6 +30,10 @@
 #error "*** Sorry, this driver requires a 2.6 kernel"
 #endif
 
+#ifdef __alpha__
+#include <asm/current.h>
+#endif
+
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
 #include <linux/autoconf.h>
 #endif
diff -aur lirc-0.8.7.orig/drivers/lirc_it87/lirc_it87.c lirc-0.8.7/drivers/lirc_it87/lirc_it87.c
--- lirc-0.8.7.orig/drivers/lirc_it87/lirc_it87.c	2010-08-16 22:20:47.000000000 +0200
+++ lirc-0.8.7/drivers/lirc_it87/lirc_it87.c	2010-09-21 12:53:12.000000000 +0200
@@ -37,6 +37,10 @@
 #include <config.h>
 #endif
 
+#ifdef __alpha__
+#include <asm/current.h>
+#endif
+
 #include <linux/version.h>
 #include <linux/module.h>
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
diff -aur lirc-0.8.7.orig/drivers/lirc_ite8709/lirc_ite8709.c lirc-0.8.7/drivers/lirc_ite8709/lirc_ite8709.c
--- lirc-0.8.7.orig/drivers/lirc_ite8709/lirc_ite8709.c	2010-01-13 20:56:13.000000000 +0100
+++ lirc-0.8.7/drivers/lirc_ite8709/lirc_ite8709.c	2010-09-21 12:53:12.000000000 +0200
@@ -19,6 +19,10 @@
  * USA
  */
 
+#ifdef __alpha__
+#include <asm/current.h>
+#endif
+
 #include <linux/version.h>
 #include <linux/module.h>
 #include <linux/interrupt.h>
diff -aur lirc-0.8.7.orig/drivers/lirc_mceusb/lirc_mceusb.c lirc-0.8.7/drivers/lirc_mceusb/lirc_mceusb.c
--- lirc-0.8.7.orig/drivers/lirc_mceusb/lirc_mceusb.c	2010-09-02 01:34:00.000000000 +0200
+++ lirc-0.8.7/drivers/lirc_mceusb/lirc_mceusb.c	2010-09-21 13:06:04.000000000 +0200
@@ -47,6 +47,11 @@
 #error "Sorry, this driver needs kernel version 2.6.5 or higher"
 #error "*******************************************************"
 #endif
+
+#ifdef __alpha__
+#include <asm/current.h>
+#endif
+
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
 #include <linux/autoconf.h>
 #endif
diff -aur lirc-0.8.7.orig/drivers/lirc_parallel/lirc_parallel.c lirc-0.8.7/drivers/lirc_parallel/lirc_parallel.c
--- lirc-0.8.7.orig/drivers/lirc_parallel/lirc_parallel.c	2010-08-27 07:06:06.000000000 +0200
+++ lirc-0.8.7/drivers/lirc_parallel/lirc_parallel.c	2010-09-21 12:53:12.000000000 +0200
@@ -38,6 +38,10 @@
 #include <linux/autoconf.h>
 #endif
 
+#ifdef __alpha__
+#include <asm/current.h>
+#endif
+
 #include <linux/module.h>
 #include <linux/sched.h>
 #include <linux/errno.h>
diff -aur lirc-0.8.7.orig/drivers/lirc_sasem/lirc_sasem.c lirc-0.8.7/drivers/lirc_sasem/lirc_sasem.c
--- lirc-0.8.7.orig/drivers/lirc_sasem/lirc_sasem.c	2010-08-16 22:20:47.000000000 +0200
+++ lirc-0.8.7/drivers/lirc_sasem/lirc_sasem.c	2010-09-21 13:04:03.000000000 +0200
@@ -41,6 +41,10 @@
 #error "*** Sorry, this driver requires kernel version 2.4.22 or higher"
 #endif
 
+#ifdef __alpha__
+#include <asm/current.h>
+#endif
+
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
 #include <linux/autoconf.h>
 #endif
diff -aur lirc-0.8.7.orig/drivers/lirc_serial/lirc_serial.c lirc-0.8.7/drivers/lirc_serial/lirc_serial.c
--- lirc-0.8.7.orig/drivers/lirc_serial/lirc_serial.c	2010-08-16 22:20:47.000000000 +0200
+++ lirc-0.8.7/drivers/lirc_serial/lirc_serial.c	2010-09-21 13:01:49.000000000 +0200
@@ -60,6 +60,10 @@
 #error "**********************************************************"
 #endif
 
+#ifdef __alpha__
+#include <asm/current.h>
+#endif
+
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
 #include <linux/autoconf.h>
 #endif
diff -aur lirc-0.8.7.orig/drivers/lirc_sir/lirc_sir.c lirc-0.8.7/drivers/lirc_sir/lirc_sir.c
--- lirc-0.8.7.orig/drivers/lirc_sir/lirc_sir.c	2010-08-16 22:20:48.000000000 +0200
+++ lirc-0.8.7/drivers/lirc_sir/lirc_sir.c	2010-09-21 13:01:09.000000000 +0200
@@ -41,6 +41,10 @@
 #endif
 #include <linux/module.h>
 
+#ifdef __alpha__
+#include <asm/current.h>
+#endif
+
 #ifdef HAVE_CONFIG_H
 # include <config.h>
 #endif
diff -aur lirc-0.8.7.orig/drivers/lirc_streamzap/lirc_streamzap.c lirc-0.8.7/drivers/lirc_streamzap/lirc_streamzap.c
--- lirc-0.8.7.orig/drivers/lirc_streamzap/lirc_streamzap.c	2010-08-16 22:20:48.000000000 +0200
+++ lirc-0.8.7/drivers/lirc_streamzap/lirc_streamzap.c	2010-09-21 13:05:08.000000000 +0200
@@ -35,6 +35,11 @@
 #error "Sorry, this driver needs kernel version 2.4.0 or higher"
 #error "*******************************************************"
 #endif
+
+#ifdef __alpha__
+#include <asm/current.h>
+#endif
+
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
 #include <linux/autoconf.h>
 #endif
diff -aur lirc-0.8.7.orig/drivers/lirc_ttusbir/lirc_ttusbir.c lirc-0.8.7/drivers/lirc_ttusbir/lirc_ttusbir.c
--- lirc-0.8.7.orig/drivers/lirc_ttusbir/lirc_ttusbir.c	2009-03-15 10:34:01.000000000 +0100
+++ lirc-0.8.7/drivers/lirc_ttusbir/lirc_ttusbir.c	2010-09-21 12:53:12.000000000 +0200
@@ -30,6 +30,10 @@
  *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
  */
 
+#ifdef __alpha__
+#include <asm/current.h>
+#endif
+
 #include <linux/version.h>
 #include <linux/kernel.h>
 #include <linux/init.h>
diff -aur lirc-0.8.7.orig/drivers/lirc_wpc8769l/lirc_wpc8769l.c lirc-0.8.7/drivers/lirc_wpc8769l/lirc_wpc8769l.c
--- lirc-0.8.7.orig/drivers/lirc_wpc8769l/lirc_wpc8769l.c	2010-03-17 15:16:16.000000000 +0100
+++ lirc-0.8.7/drivers/lirc_wpc8769l/lirc_wpc8769l.c	2010-09-21 13:05:46.000000000 +0200
@@ -37,6 +37,10 @@
 #error "**********************************************************"
 #endif
 
+#ifdef __alpha__
+#include <asm/current.h>
+#endif
+
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
 #include <linux/autoconf.h>
 #endif
