--- lirc-0.10.2/tools/Makefile.am.orig	2022-10-08 10:26:12.255237014 +0200
+++ lirc-0.10.2/tools/Makefile.am	2022-10-08 10:27:13.968584868 +0200
@@ -100,21 +100,8 @@
 	$(SED) -i -e 's/@version@/$(VERSION)/' \
 	    $(DESTDIR)$(bindir)/lirc-config-tool \
 	    $(DESTDIR)$(bindir)/irdb-get
-	cd $(DESTDIR)$(bindir);  \
-	    $(SED) -i '\|^#!/usr/bin/env|s|python3|$(PYTHON)|' \
-	    irdb-get \
-	    irtext2udp \
-	    lirc-init-db \
-	    lirc-postinstall \
-	    pronto2lirc
-	cd $(DESTDIR)$(sbindir);  \
-	    $(SED) -i '\|^#!/usr/bin/env|s|python3|$(PYTHON)|' lircd-setup
-
 
 install-data-hook:
-	cd $(DESTDIR)/$(setupdir);  \
-	    $(SED) -i '1s|python3[0-9\.]*|$(PYTHON)|' \
-	        $(notdir $(dist_setup_SCRIPTS))
 	$(PYTHON) $(srcdir)/make_rel_symlink.py \
 	    $(DESTDIR)/$(setupdir)/lirc-setup \
 	    $(DESTDIR)/$(bindir)/lirc-setup
